[pytree] Add optimized `tree_leaves` implementation

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112324
* __->__ #112323
* #112327

pytree is used in many hot paths for dynamo tracing and in many cases we don't
care about the tree spec and just want the flattened list. This improves
`pytree.tree_leaves` to not construct the spec which gives a noticeable
performance improvement when multiplied by the many times it gets called
during tracing.

Concretely, I see a 2x speedup compared to `tree_flatten` in this benchmark:
```python
import torch.utils._pytree as pytree
%timeit pytree.tree_flatten([((100, 100), (100, 1)), dict(device="cuda")])[0]
%timeit pytree.tree_leaves([((100, 100), (100, 1)), dict(device="cuda")])
```

cc @zou3519

peterbell10 committed:
[pytree] Add optimized `tree_leaves` implementation

pytree is used in many hot paths for dynamo tracing and in many cases we don't
care about the tree spec and just want the flattened list. This improves
`pytree.tree_leaves` to not construct the spec which gives a noticeable
performance improvement when multiplies by the millions of times it gets called
during tracing.

[ghstack-poisoned]

peterbell10 committed:
Update on "[pytree] Add optimized `tree_leaves` implementation"


pytree is used in many hot paths for dynamo tracing and in many cases we don't
care about the tree spec and just want the flattened list. This improves
`pytree.tree_leaves` to not construct the spec which gives a noticeable
performance improvement when multiplied by the many times it gets called
during tracing.

Concretely, I see a 2x speedup compared to `tree_flatten` in this benchmark:
```python
import torch.utils._pytree as pytree
%timeit pytree.tree_flatten([((100, 100), (100, 1)), dict(device="cuda")])[0]
%timeit pytree.tree_leaves([((100, 100), (100, 1)), dict(device="cuda")])
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[pytree] Add optimized `tree_leaves` implementation"


pytree is used in many hot paths for dynamo tracing and in many cases we don't
care about the tree spec and just want the flattened list. This improves
`pytree.tree_leaves` to not construct the spec which gives a noticeable
performance improvement when multiplied by the many times it gets called
during tracing.

Concretely, I see a 2x speedup compared to `tree_flatten` in this benchmark:
```python
import torch.utils._pytree as pytree
%timeit pytree.tree_flatten([((100, 100), (100, 1)), dict(device="cuda")])[0]
%timeit pytree.tree_leaves([((100, 100), (100, 1)), dict(device="cuda")])
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[pytree] Add optimized `tree_leaves` implementation"


pytree is used in many hot paths for dynamo tracing and in many cases we don't
care about the tree spec and just want the flattened list. This improves
`pytree.tree_leaves` to not construct the spec which gives a noticeable
performance improvement when multiplied by the many times it gets called
during tracing.

Concretely, I see a 2x speedup compared to `tree_flatten` in this benchmark:
```python
import torch.utils._pytree as pytree
%timeit pytree.tree_flatten([((100, 100), (100, 1)), dict(device="cuda")])[0]
%timeit pytree.tree_leaves([((100, 100), (100, 1)), dict(device="cuda")])
```

[ghstack-poisoned]

lezcano commented:
cc @XuehaiPan in case you want to implement this in optree as well

lezcano reviewed (approved):
~~LGTM, but it'd be good to add the optimisations from @XuehaiPan as well~~ Wrong pr

XuehaiPan commented:
> in case you want to implement this in optree as well

@lezcano I'll benchmark the overhead of creating the `treespec` object on the C++ side. I think it might be marginal.

XuehaiPan commented on torch/utils/_pytree.py:
```diff
@@ -274,7 +274,19 @@ def tree_flatten(pytree: PyTree) -> Tuple[List[Any], TreeSpec]:
 
 def tree_leaves(pytree: PyTree) -> List[Any]:
     """Get a list of leaves of a pytree."""
-    return tree_flatten(pytree)[0]
+    if _is_leaf(pytree):
+        return [pytree]
+
+    node_type = _get_node_type(pytree)
+    flatten_fn = SUPPORTED_NODES[node_type].flatten_fn
+    child_pytrees, context = flatten_fn(pytree)
+
+    # Recursively flatten the children
+    result : List[Any] = []
+    for child in child_pytrees:
+        result += tree_leaves(child)
+
+    return result
```
I think we could have an even further optimized implementation using generators. I haven't profiled this yet. The implementation in this PR will create intermediate `list` objects in the recursive `tree_leaves` call.

peterbell10 replied:
Yeah you don't even need a generator, we can pass the result as an argument to a helper and append directly into the list.
```python
def tree_leaves_helper(pytree, result):
    if _isleaf(pytree):
        result.append(pytree)
    # [snip]
    for child in child_pytrees:
        tree_leaves_helper(child, result)
```

This way there are no intermediates at all. In my small benchmark it was only a 5% uplift, and I didn't want to diverge too far from the `tree_flatten` implementation so they would be easier to keep in sync. Could maybe change both though in a further PR.

jon-chuang commented on torch/utils/_pytree.py:
```diff
@@ -274,7 +274,19 @@ def tree_flatten(pytree: PyTree) -> Tuple[List[Any], TreeSpec]:
+
+    # Recursively flatten the children
+    result : List[Any] = []
+    for child in child_pytrees:
```
nit: maybe it's better to have an internal `_tree_flatten` function that optionally computes tree spec. Then wrap it in the public APIs.

Likewise for `tree_structure`

peterbell10 replied:
I tried this and every way I tried made `tree_leaves` slower and none made `tree_structure` significantly faster so I think this is a reasonable position.

peterbell10 committed:
Update on "[pytree] Add optimized `tree_leaves` implementation"


pytree is used in many hot paths for dynamo tracing and in many cases we don't
care about the tree spec and just want the flattened list. This improves
`pytree.tree_leaves` to not construct the spec which gives a noticeable
performance improvement when multiplied by the many times it gets called
during tracing.

Concretely, I see a 2x speedup compared to `tree_flatten` in this benchmark:
```python
import torch.utils._pytree as pytree
%timeit pytree.tree_flatten([((100, 100), (100, 1)), dict(device="cuda")])[0]
%timeit pytree.tree_leaves([((100, 100), (100, 1)), dict(device="cuda")])
```

cc zou3519

[ghstack-poisoned]

peterbell10 committed:
Update on "[pytree] Add optimized `tree_leaves` implementation"


pytree is used in many hot paths for dynamo tracing and in many cases we don't
care about the tree spec and just want the flattened list. This improves
`pytree.tree_leaves` to not construct the spec which gives a noticeable
performance improvement when multiplied by the many times it gets called
during tracing.

Concretely, I see a 2x speedup compared to `tree_flatten` in this benchmark:
```python
import torch.utils._pytree as pytree
%timeit pytree.tree_flatten([((100, 100), (100, 1)), dict(device="cuda")])[0]
%timeit pytree.tree_leaves([((100, 100), (100, 1)), dict(device="cuda")])
```

cc zou3519

[ghstack-poisoned]
